egs/librispeech/asr/simple_v1/mmi_mbr_train.py

```diff
@@ -0,0 +1,508 @@
+    else:
+        with torch.no_grad():
+            num, den, decoding_graph = graph_compiler.compile(texts, P)
+
+    assert num.requires_grad == is_training
+    assert den.requires_grad is False
+    assert decoding_graph.requires_grad is False
+
+    num = num.to(device)
+    den = den.to(device)
+
+    decoding_graph = decoding_graph.to(device)
+
+    dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+    assert nnet_output.device == device
+
+    logging.info('intersect dense num')
+    num_graph = k2.intersect_dense(num, dense_fsa_vec, 10.0, seqframe_idx_name='seqframe_idx')
+
+    logging.info('intersect dense den')
+    den_graph = k2.intersect_dense(den, dense_fsa_vec, 10.0)
+
+    logging.info('num get tot scores')
+    num_tot_scores = num_graph.get_tot_scores(log_semiring=True,
+                                              use_double_scores=True)
+
+    logging.info('den get tot scores')
+    den_tot_scores = den_graph.get_tot_scores(log_semiring=True,
+                                              use_double_scores=True)
+
+    tot_scores = num_tot_scores - den_scale * den_tot_scores
+
+    (tot_score, tot_frames,
+     all_frames) = get_tot_objf_and_num_frames(tot_scores,
+                                               supervision_segments[:, 2])
+
+    logging.info('Decoding. It will not terminate. Debug it! Use kill -9 to kill it')
+
+    decoding_lattice = k2.intersect_dense_pruned(
+        decoding_graph,
+        dense_fsa_vec,
+        7.0,
+        3.0,
+        30,
+        100,
+        seqframe_idx_name='seqframe_idx')
+
+    num_rows = dense_fsa_vec.scores.shape[0]
+    num_cols = dense_fsa_vec.scores.shape[1] - 1
+    logging.info('mbr num create sparse')
+    mbr_num_sparse = k2.create_sparse(rows=num_graph.seqframe_idx,
+                                      cols=num_graph.phones,
+                                      values=num_graph.get_arc_post(
+                                          True, True).exp(),
+                                      size=(num_rows, num_cols),
+                                      min_col_index=0)
+
+    logging.info('mbr lattice create sparse')
+    mbr_den_sparse = k2.create_sparse(rows=decoding_lattice.seqframe_idx,
+                                      cols=decoding_lattice.phones,
+                                      values=decoding_lattice.get_arc_post(
+                                          True, True).exp(),
+                                      size=(num_rows, num_cols),
+                                      min_col_index=0)
+    # NOTE: Due to limited support of PyTorch's autograd for sparse tensors,
+    # we cannot use (mbr_num_sparse - mbr_den_sparse) here
+    #
+    # The following works only for torch >= 1.7.0
+    logging.info('mbr sum')
+    mbr_loss = torch.sparse.sum(
+        k2.sparse.abs((mbr_num_sparse + (-mbr_den_sparse)).coalesce()))
+
+    mmi_loss = -tot_score
+
+    total_loss = mmi_loss + mbr_loss
```

Comment:
Here is the total loss.

Reply:
OK..

Reply:
Would be nice if you can verify that if you make this `total_loss = mmi_loss + 0.5 * mbr_loss`, the MMI objf gets better and the MBR objf gets worse (in the plots).

Reply:
Will verify it before merging.

Reply:
> the MMI objf gets better and the MBR objf gets worse (in the plots).

Both losses are decreasing. MBR losses do not get worse.
Please see https://tensorboard.dev/experiment/QFeH3h7qTgOePVhWoxKi8Q/#scalars&_smoothingWeight=0

Does this reflect any problems?

Reply:
I just mean I'd expect that the MBR loss would not reach quite as good as value as when the weight is 1.0, and the MMI loss would reach a slightly better value.

Reply:
I see. Will report the final WER after 10 epochs tomorrow.